internal/observability/collector/collector.go

```diff
@@ -0,0 +1,139 @@
+//
+// Copyright (C) 2019-2020 Vdaas.org Vald team ( kpango, rinx, kmrmt )
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package collector provides metrics collector
+package collector
+
+import (
+	"context"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/vdaas/vald/internal/errgroup"
+	"github.com/vdaas/vald/internal/observability/metrics"
+	"github.com/vdaas/vald/internal/safety"
+)
+
+var (
+	initMetrics []metrics.Metric
+	instance    *collector
+	once        sync.Once
+)
+
+type Collector interface {
+	PreStart(ctx context.Context) error
+	Start(ctx context.Context) <-chan error
+	Stop(ctx context.Context)
+}
+
+type collector struct {
+	duration time.Duration
+	metrics  []metrics.Metric
+	eg       errgroup.Group
+}
+
+func New(opts ...CollectorOption) (Collector, error) {
+	co := new(collector)
+
+	for _, opt := range append(collectorDefaultOpts, opts...) {
+		err := opt(co)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	co.eg = errgroup.Get()
+
+	return co, nil
+}
+
+func Register(ms ...metrics.Metric) error {
```

Comment:
did you think about another way to set collector instead of static method

Reply:
I implemented it before merging #231. Before it, there's no way to fetch NGT metrics except for registering its metric from pkg/agent/ngt/service.
But now, we can inject NGT metrics at initialize time. so i'll refactor it. :muscle: